clang/docs/SYCLSupport.rst

```diff
@@ -1,119 +1,542 @@
+kernel invocation.
+This is accomplished by, in the definition of a SYCL kernel invocation function,
+including a call to a function declared with the attribute.
+For example:
+
+.. code-block:: C++
 
-     pointer_t m_Pointer;
+   namespace sycl {
+     class handler {
+       template <typename KernelName, typename KernelType>
+       [[clang::sycl_kernel_entry_point(KernelName)]]
+       void kernel_entry_point(KernelType kernelFunc) {
+         kernelFunc();
+       }
      public:
-     pointer_t get() { return m_Pointer; }
-     T& operator* () { return *reinterpret_cast<T*>(m_Pointer); }
+       template <typename KernelName, typename KernelType>
+       void single_task(const KernelType &kernelFunc) {
+         kernel_entry_point<KernelName>(kernelFunc);
+       }
+     };
    }
-   #else
-   // CPU/host implementation
-   template <typename T, address_space AS> class multi_ptr {
-     T *m_Pointer; // regular undecorated pointer
-     public:
-     T *get() { return m_Pointer; }
-     T& operator* () { return *m_Pointer; }
+
+The arguments of the call to ``kernel_entry_point()`` constitute the parameters
+of a SYCL kernel.
+The body of the ``kernel_entry_point()`` function contains the statements
+required to execute the SYCL kernel (e.g., an invocation of the call operator of
+the SYCL kernel object as in this example).
+The call to ``kernel_entry_point()`` in ``single_task()`` establishes a common
+point of SYCL kernel invocation for Clang and the SYCL run-time library.
+With that point established, the tasks required to actually execute a SYCL
+kernel are delegated according to the following division of responsibilities.
+
+Clang is responsible for:
+
+- Validating that all kernel argument types (e.g., the deduced parameter types
+  of the ``kernel_entry_point()`` function above) satisfy the requirements
+  specified in
+  `SYCL 2020 section 4.12.4, "Rules for parameter passing to kernels" <SYCL-2020-4.12.4_>`_.
+- Informing the SYCL run-time library of the presence of subobjects of SYCL
+  types that require special handling within kernel arguments.
+- Generating an offload kernel entry point function for each SYCL kernel for
+  each target device, generating a name for it derived from the SYCL kernel
+  name, and informing the SYCL run-time library of the generated name.
+
+The SYCL run-time library is responsible for:
+
+- Selecting a device on which to execute the kernel.
+- Copying the SYCL kernel object and any other kernel arguments to the device.
+- Informing Clang of additional parameters required for the offload kernel
+  entry point based on the presence of subobjects of SYCL types that require
+  special handling within kernel arguments.
+- Scheduling execution of the offload kernel entry point function on the
+  selected device.
+
+The SYCL run-time library tasks are expected to be performed in conjunction
+with an offload backend such as liboffload, OpenCL, CUDA, Hip, or Level Zero;
```

Comment:
```suggestion
with an offload backend such as liboffload, OpenCL, CUDA, HIP, or Level Zero;
```

Reply:
Thank you, fixed in https://github.com/llvm/llvm-project/pull/170602/commits/4f3a5cdba2fa9938ab4d67e41bf752155fb71c8b.